### PR TITLE
DM-30977: Accumulate log messages per quantum and store in butler

### PR DESCRIPTION
**Pull request by timj**
Relies on lsst/daf_butler#546 and lsst/pipe_base#193

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: python/lsst/ctrl/mpexec/singleQuantumExecutor.py [hunk 1]
```diff
@@ -163,6 +182,24 @@ def setupLogging(self, taskDef, quantum):
 
         ButlerMDC.MDC("LABEL", label)
 
+        # Add the handler to the root logger.
+        # How does it get removed reliably?
```
Comment (andy-slac) [inline python/lsst/ctrl/mpexec/singleQuantumExecutor.py:186]: Do you mean if writeLogRecords is not called?

File: python/lsst/ctrl/mpexec/singleQuantumExecutor.py [hunk 2]
```diff
@@ -374,6 +411,50 @@ def writeMetadata(self, quantum, metadata, taskDef, butler):
...
+                    f" and execution") from exc
+
+            if isinstance(self.log_handler, ButlerLogRecordHandler):
+                butler.put(self.log_handler.records, ref[0])
+
+                # Clear the records in case the handler is reused.
+                self.log_handler.records.clear()
+            else:
+                if filename is None:
+                    raise RuntimeError("Something bad went wrong extracting filename from"
```
Comment (andy-slac) [inline python/lsst/ctrl/mpexec/singleQuantumExecutor.py:443]: `Something bad went wrong` - does it mean it's not bad anymore? 😄

**Review by andy-slac — APPROVED** [review]
Looks good

File: python/lsst/ctrl/mpexec/singleQuantumExecutor.py (hunk 1 above)
Comment (timj) [reply python/lsst/ctrl/mpexec/singleQuantumExecutor.py:195]: Yes, maybe I should wrap the entire `execute` method in a `try/finally`.  On the other hand if I do fail to remove the handler eventually you'll get a `ResourceWarning` -- the tests did actually fail until I fixed some of the edge cases for calling writeLogRecords

File: python/lsst/ctrl/mpexec/singleQuantumExecutor.py (hunk 2 above)
Comment (timj) [reply python/lsst/ctrl/mpexec/singleQuantumExecutor.py:453]: Maybe it should be an assert here because this check is never meant to trigger -- if it does something unexpected happened.

File: python/lsst/ctrl/mpexec/singleQuantumExecutor.py (hunk 1 above)
Comment (andy-slac) [reply python/lsst/ctrl/mpexec/singleQuantumExecutor.py:195]: Or maybe make `setupLogging` to return a context manager and wrap all code in `with`?

Comment (timj) [reply python/lsst/ctrl/mpexec/singleQuantumExecutor.py:195]: Hmm. Very good point.

Comment (timj) [reply python/lsst/ctrl/mpexec/singleQuantumExecutor.py:200]: I also took the opportunity to add a context manager for the ButlerMDC setting so that it gets reset on exit.  Feel free to take a second look at this PR -- it's going to be a few hours before Jenkins completes.

File: python/lsst/ctrl/mpexec/singleQuantumExecutor.py [hunk 3]
```diff
@@ -77,92 +90,141 @@ class SingleQuantumExecutor(QuantumExecutor):
...
+
+        Notes
+        -----
+        Expected to be used as a context manager to ensure that logging
+        records are inserted into the butler once the quantum has been
+        executed:
+
+        .. code-block:: py
+
+           with self.captureLogging(taskDef, quantum):
```
Comment (andy-slac) [inline python/lsst/ctrl/mpexec/singleQuantumExecutor.py:192]: Needs a `butler` argument too

**Review by andy-slac — APPROVED** [review]
Looks good, one docstring comment.
